src/Microsoft.AspNet.Mvc.ViewFeatures/ViewFeatures/DefaultHtmlGenerator.cs

```diff
@@ -1154,12 +1157,52 @@ internal static object GetModelStateValue(ViewContext viewContext, string key, T
                     nameof(expression));
             }
 
+            var inputTypeString = GetInputTypeString(inputType);
             var tagBuilder = new TagBuilder("input");
             tagBuilder.TagRenderMode = TagRenderMode.SelfClosing;
             tagBuilder.MergeAttributes(htmlAttributes);
-            tagBuilder.MergeAttribute("type", GetInputTypeString(inputType));
+            tagBuilder.MergeAttribute("type", inputTypeString);
             tagBuilder.MergeAttribute("name", fullName, replaceExisting: true);
 
+            switch (inputType)
+            {
+                case InputType.Text:
+                    object suppliedTypeString;
+                    if (htmlAttributes != null &&
+                        htmlAttributes.TryGetValue("type", out suppliedTypeString) &&
+                        suppliedTypeString != null)
+                    {
+                        if (_placeholderInputTypes.Contains(suppliedTypeString))
+                        {
+                            goto case InputType.Password;
+                        }
+                    }
+                    else if (inputTypeString == "text")
+                    {
+                        goto case InputType.Password;
```

Comment:
Very true. I want to say was trying to avoid declaring a local but with all I've learned in the past week about MSIL I can now see why that was silly.

- Derek Gray

> On Jan 25, 2016, at 2:56 PM, Ryan Nowak <notifications@github.com> wrote:
> 
> In src/Microsoft.AspNet.Mvc.ViewFeatures/ViewFeatures/DefaultHtmlGenerator.cs:
> 
> > +            switch (inputType)
> > +            {
> > +                case InputType.Text:
> > +                    object suppliedTypeString;
> > +                    if (htmlAttributes != null &&
> > +                        htmlAttributes.TryGetValue("type", out suppliedTypeString) &&
> > +                        suppliedTypeString != null)
> > +                    {
> > +                        if (_placeholderInputTypes.Contains(suppliedTypeString))
> > +                        {
> > +                            goto case InputType.Password;
> > +                        }
> > +                    }
> > +                    else if (inputTypeString == "text")
> > +                    {
> > +                        goto case InputType.Password;
> This is a really strange way to formulate this logic. It looks like this is in the middle of a refactor and not quite finished.
> 
> I'd expect to see something like:
> 
> var canIncludePlaceholder false;
> 
> switch
> {
> case InputType.Text:
>   if (...) // conditions for text
>   {
>     canIncludePlaceholder = true;
>   }
>   break;
> case InputType.Text:
>   if (...) // conditions for password
>   {
>     canIncludePlaceholder = true;
>   }
>   break;
> }
> 
> if (canIncludePlaceholder)
> {
>    // Add the placeholder
> }
> —
> Reply to this email directly or view it on GitHub.
>